controllers/ibmpowervsimage_controller.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"context"
+
+	"github.com/IBM-Cloud/power-go-client/power/models"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/util"
+)
+
+// IBMPowerVSImageReconciler reconciles a IBMPowerVSImage object
+type IBMPowerVSImageReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=ibmpowervsimages,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=ibmpowervsimages/status,verbs=get;update;patch
+
+// Reconcile implements controller runtime Reconciler interface and handles reconcileation logic for IBMPowerVSImage.
+func (r *IBMPowerVSImageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := r.Log.WithValues("ibmpowervsimage", req.NamespacedName)
+
+	ibmImage := &v1beta1.IBMPowerVSImage{}
+	err := r.Get(ctx, req.NamespacedName, ibmImage)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if err := ctrl.SetControllerReference(ibmImage, ibmImage, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the Image.
+	image, err := util.GetOwnerImage(ctx, r.Client, ibmImage.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if image == nil {
+		log.Info("Image Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	// Create the scope.
+	imageScope, err := scope.NewPowerVSImageScope(scope.PowerVSImageScopeParams{
+		Client:          r.Client,
+		Logger:          log,
+		IBMPowerVSImage: ibmImage,
+	})
+
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := imageScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !ibmImage.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(imageScope)
+	}
+
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+	return r.reconcile(ctx, imageScope)
+}
+
+func (r *IBMPowerVSImageReconciler) reconcile(ctx context.Context, imageScope *scope.PowerVSImageScope) (ctrl.Result, error) {
+	controllerutil.AddFinalizer(imageScope.IBMPowerVSImage, v1beta1.IBMPowerVSImageFinalizer)
+
+	lastJob, err := imageScope.GetImportJob()
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if lastJob != nil {
+		if *lastJob.Status.State != "completed" {
+			return ctrl.Result{}, nil
```

Comment:
what happens when there are more than one image/job present?

Reply:
It will be good if we can track the `jobID` from the create call(store that in the status of the image and then compare/track that here will be better approach.

Reply:
@mkumatag I think there is a limit on import job on PowerVS service instance (it is restricted to 1 job at a time), so above logic should be good considering that, if that is not the case then we can switch to the JobID approach.

Reply:
Its always good to capture the `jobID` and compare the result with a current running job, there is a possibility that the current import has already finished and some other import started meanwhile then we will be end up waiting for the wrong job.